cuML/src/common/cumlHandle.cpp

```diff
@@ -0,0 +1,435 @@
+
+cumlHandle_impl::cumlHandle_impl()
+    : _dev_ids(1,0), _deviceAllocator( std::make_shared<defaultDeviceAllocator>() ), _hostAllocator( std::make_shared<defaultHostAllocator>() )
+{
+    createResources();
+}
+
+cumlHandle_impl::~cumlHandle_impl()
+{
+    destroyResources();
+}
+
+void cumlHandle_impl::setDevice( int dev_id )
+{
+    destroyResources();
+    _dev_ids[0] = dev_id;
+    createResources();
+}
+
+void cumlHandle_impl::setDevices(const std::vector<int>& dev_ids )
+{
+    destroyResources();
+    _dev_ids = dev_ids;
+    createResources();
+}
+
+int cumlHandle_impl::getDevice( int dev_idx ) const
+{
+    return _dev_ids[dev_idx];
+}
+
+int cumlHandle_impl::getNumDevices() const
+{
+    return _dev_ids.size();
+}
+
+void cumlHandle_impl::setStream( cudaStream_t stream )
+{
+    _userStream = stream;
+}
+
+cudaStream_t cumlHandle_impl::getStream() const
+{
+    return _userStream;
+}
+
+cudaStream_t cumlHandle_impl::getDeviceStream( int dev_idx ) const
+{
+    return _streams[dev_idx*_num_streams + 0];
+}
+
+void cumlHandle_impl::setDeviceAllocator( std::shared_ptr<deviceAllocator> allocator )
+{
+    _deviceAllocator = allocator;
+}
+
+std::shared_ptr<deviceAllocator> cumlHandle_impl::getDeviceAllocator() const
+{
+    return _deviceAllocator;
+}
+
+void cumlHandle_impl::setHostAllocator( std::shared_ptr<hostAllocator> allocator )
+{
+    _hostAllocator = allocator;
+}
+
+std::shared_ptr<hostAllocator> cumlHandle_impl::getHostAllocator() const
+{
+    return _hostAllocator;
+}
+
+cublasHandle_t cumlHandle_impl::getCublasHandle( int dev_idx ) const
+{
+    return _cublas_handles[dev_idx];
+}
+
+cusolverDnHandle_t cumlHandle_impl::getcusolverDnHandle( int dev_idx ) const
+{
+    return _cusolverDn_handles[dev_idx];
+}
+
+cusparseHandle_t cumlHandle_impl::getcusparseHandle( int dev_idx ) const
+{
+    return _cusparse_handles[dev_idx];
+}
+
+cudaStream_t cumlHandle_impl::getInternalStream( int sid ) const
+{
+    return _streams[sid];
+}
+
+int cumlHandle_impl::getNumInternalStreams() const
+{
+    return _num_streams;
+}
+
+void cumlHandle_impl::waitOnUserStream() const
+{
+    CUDA_CHECK( cudaEventRecord( _event, _userStream ) );
+    for (auto s : _streams)
+    {
+        CUDA_CHECK( cudaStreamWaitEvent( s, _event, 0 ) );
+    }
+}
+
+void cumlHandle_impl::waitOnInternalStreams() const
+{
+    for (auto s : _streams) 
+    {
+        CUDA_CHECK( cudaEventRecord( _event, s ) );
+        CUDA_CHECK( cudaStreamWaitEvent( _userStream, _event, 0 ) );
+    }
+}
+
+void cumlHandle_impl::createResources()
+{
+    CUDA_CHECK( cudaEventCreateWithFlags( &_event, cudaEventDisableTiming ) );
+    cudaDeviceScope _( 0 );
+    for (auto dev_id: _dev_ids) 
+    {
+        CUDA_CHECK( cudaSetDevice(dev_id) );
+
+        cudaStream_t stream;
+        CUDA_CHECK( cudaStreamCreate(&stream) );
+
+        cublasHandle_t cublas_handle;
+        CUBLAS_CHECK( cublasCreate(&cublas_handle) );
```

Comment:
Are you trying to avoid runtime overhead by pre-creating a handle to each LINALG library at startup? What is the cost of creation? What if the user doesn't ever use, say cuSolver? Does creating the handle cost anything in terms of memory usage?  Is it not acceptable to lazily create the handles, e.g. when the user first wants to use the library? (I know, implementation detail. Just food for thought, I don't need answers.)

Reply:
Yes the assumption is that usually with one handle one or multiple ML algorithms are called multiple times. So creating the library handle in `cumlHandle_impl` avoids creation and recreation. You have a good point regarding lazily creating the library handles as not all algorithms need all of them. On the other hand if they are created lazily a user of cuML does not have control when they are created. For that reason I am in favor of keeping it as it is and consider lazy initialization if a use case suffering for unnecessary initialization overhead arises. Does that make sense to you?

Reply:
Yup